src/hotspot/share/opto/vector.cpp

```diff
@@ -316,12 +318,22 @@ void PhaseVector::expand_vbox_node(VectorBoxNode* vec_box) {
 Node* PhaseVector::expand_vbox_node_helper(Node* vbox,
                                            Node* vect,
                                            const TypeInstPtr* box_type,
-                                           const TypeVect* vect_type) {
+                                           const TypeVect* vect_type,
+                                           VectorSet &visited) {
+  // JDK-8304948 shows an example that there may be a circle in the graph
+  // after the transformation of Phi:
+  // Phi (VectorBox VectorBox) => VectorBox (Phi Phi)
+  if (visited.test_set(vbox->_idx)) {
+    assert(vbox->is_Phi(), "not a phi");
```

Comment:
I think it is.

The CYCLE is derived from merge_through_phi, in which only Phi CYCLE is allowed [1].

This method `expand_vbox_node_helper` serves to locate the target node `Proj(VectorBoxAllocate)` and then replace it with a punch of other nodes. When expanding VectorBox, the normal case is `VectorBox (Proj value)`. If it is in shape of `VectorBox (Phi1 Phi2)`[2] or `VectorBox (Phi1 vect)`[3],  I suppose it should be transformed by merge_through_phi.

In `expand_vbox_node_helper`, it is recursive only at Phi1 since it is where Proj in. 

[1] https://github.com/e1iu/jdk/blob/ENTLLT-4778-external/src/hotspot/share/opto/cfgnode.cpp#L2574
[2] https://github.com/e1iu/jdk/blob/ENTLLT-4778-external/src/hotspot/share/opto/vector.cpp#L331
[3] https://github.com/e1iu/jdk/blob/ENTLLT-4778-external/src/hotspot/share/opto/vector.cpp#L341

Reply:
Okay, thanks for the details! Looks good.